### PR TITLE
Add kmeans page to cookbook

### PR DESCRIPTION
**Pull request by OXPHOS**
- There's no parameter `CLabels` in class `CKMeans`, so I can't find a way to use `apply_*`, `eval.evaluate`, so as to compare test and training dataset.
- Though I didn't see why `CKMeans` cannot have `CLabels` - we can just label the clusters 1..N.

- I thought about evaluating the clustering performance by calculating the Euclidean distances between centers of training dataset and test dataset, but there's no handy method for now.

- I didn't see the difference between dataset `fm_train_real.dat` and `classifier_binary__2d_linear_features_train.dat`but I think it doesn't really matter which one to use..?

### CODE REVIEW

**Comment by OXPHOS** [comment]
Some general issues - 
1. I ran into the same problem as in #3029 when I tried to run `clustering_kmeans_modular.py`. Import of `modshogun` resulted in `segmentation fault 11` and I had no idea why.

2. I was trying to translate the meta scripts to other languages. I did `./generate.py --input examples --output outputs `, but was told `ctags` cannot be NoneType. The package `ctags` can be found in cmake: `Ctags flavour: Exuberant` (?). I haven't looked into this problem yet.

**Comment by lisitsyn** [comment]
@OXPHOS I will take a look at ctags issue. This was actually caused by me.

**Comment by Saurabh7** [comment]
@OXPHOS 

- Have you tried `apply_multiclass()` , should work for KMeans i think. https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/machine/DistanceMachine.cpp#L208

- About calculating the distances to closest centers, I thought it should be possible too. I was actually planning to add it after PR  https://github.com/shogun-toolbox/shogun/pull/3166

**Comment by OXPHOS** [comment]
@lisitsyn  Hey the error message is interestingly different from the last time I submitted but I guess it still has something to do with `ctags`...? I cannot make meta-examples at all now.

Please see below for details. Thanks!

```
Built target ctags
Generating examples from meta-language
Traceback (most recent call last):
  File "/Users/zora/Github/shogun/examples/meta/generator/generate.py", line 118, in <module>
    includedTargets=args.targets, storeVars=storeVars)
  File "/Users/zora/Github/shogun/examples/meta/generator/generate.py", line 58, in translateExamples
    storeVars=storeVars if target['FileExtension']==".cpp" else False)
  File "/Users/zora/Github/shogun/examples/meta/generator/translate.py", line 365, in translate
    return translator.translateProgram(ast["Program"], programName, tags, storeVars)
  File "/Users/zora/Github/shogun/examples/meta/generator/translate.py", line 56, in translateProgram
    dependencies=self.dependenciesString(),
  File "/Users/zora/Github/shogun/examples/meta/generator/translate.py", line 121, in dependenciesString
    allClassDependencies = template.substitute(classlist=self.seperatedClassDependencies("AllClasses"))
  File "/Users/zora/Github/shogun/examples/meta/generator/translate.py", line 159, in seperatedClassDependencies
    csdependencies += elementTemplate.substitute(element=x, include=self.getIncludePathForClass(x))
  File "/Users/zora/Github/shogun/examples/meta/generator/translate.py", line 181, in getIncludePathForClass
    raise Exception('Failed to obtain include path for %s' % (' or '.join(variants)))
Exception: Failed to obtain include path for int32_t or Cint32_t or int32_t or Cint32_t
make[3]: *** [examples/meta/CMakeFiles/meta_examples] Error 1
make[2]: *** [examples/meta/CMakeFiles/meta_examples.dir/all] Error 2
make[1]: *** [doc/cookbook/CMakeFiles/cookbook.dir/rule] Error 2
```

File: doc/cookbook/source/examples/classifier/kmeans.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
+==================
+:math:`k`-means clustering
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: Minors: Capital K please. Also, can you avoid the math in the header (only there)

File: doc/cookbook/source/examples/classifier/kmeans.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+The :math:`n` observations are represented by :math:`n` :math:`d`-dimensional real vecoters, :math:`\mathbf{x} = (x_1, x_2, ..., x_n)`.
+
+
+
+In :math:`k`-means clustering, the :math:`n` observations will partitioned into :math:`k` (:math:`\leq n`) sets :math:`\mathbf{S} = {S_1, S_2, ..., S_k}`, with minimal within-cluster sum of squares (WCSS) (sum of distance functions of each point in the cluster to the :math:`k^{th}` center). 
+In other words, its objective is to find:
+
+.. math::
+   k = \underset{\mathbf{S}}{argmin} \sum_{i=1}^{k}\sum_{\mathbf{x}\in S_k}\left \|\boldsymbol{x} - \boldsymbol{\mu}_i  \right \|^{2}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: please use \argmin, which is already defined

File: doc/cookbook/source/examples/classifier/kmeans.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
+==================
+:math:`k`-means clustering
+==================
+:math:`k`-means clustering aims to partition :math:`n` observations into :math:`k` (:math:`\leq n`) clusters in which each observation belongs to the cluster with the nearest mean, serving as a prototype of the cluster.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: no () around the <=

File: doc/cookbook/source/examples/classifier/kmeans.rst [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+The :math:`n` observations are represented by :math:`n` :math:`d`-dimensional real vecoters, :math:`\mathbf{x} = (x_1, x_2, ..., x_n)`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: typo: please run a spell checker before you send cookbook pages

File: doc/cookbook/source/examples/classifier/kmeans.rst [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+
+
+In :math:`k`-means clustering, the :math:`n` observations will partitioned into :math:`k` (:math:`\leq n`) sets :math:`\mathbf{S} = {S_1, S_2, ..., S_k}`, with minimal within-cluster sum of squares (WCSS) (sum of distance functions of each point in the cluster to the :math:`k^{th}` center).
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: Try to be a bit more concise (that is: less text) here. Just stating the objective should be fine

File: doc/cookbook/source/examples/classifier/kmeans.rst [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+.. sgexample:: kmeans.sg:create_instance
+
+Then we train the dataset:
+
+.. sgexample:: kmeans.sg:train_and_apply
+
+And we can extract centers and radius of each cluster:
+
+.. sgexample:: kmeans.sg:extract_centers_and_radius
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: This needs an illustration of
 * apply method (I guess cant be done right now as discussed in the thread)
 * evaluation (based on above)
 * How to run the mini batch version

File: doc/cookbook/source/examples/classifier/kmeans.rst (hunk 2 above)
Comment (OXPHOS) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: Do you mean `\underset{\mathbf{S}}{\argmin}` ?

The online LaTeX editor I'm using return blank for `\argmin` but `\arg\min` works...I guess it is somehow outdated?

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/kmeans.rst]: \argmin is not part of standard latex, but as mentioned in the [readme](https://github.com/shogun-toolbox/shogun/tree/develop/doc/cookbook), we have custom macros defined [here](https://github.com/shogun-toolbox/shogun/blob/develop/doc/cookbook/source/static/mathconf.js) that you can use and extend

**Comment by OXPHOS** [comment]
Error:
> 28/32 Test #29: generated_ruby-classifier-kmeans ................................***Failed    0.10 sec
> /opt/shogun/build/examples/meta/ruby/classifier/kmeans.rb:28:in `<main>': undefined method `set_mbKMeans_params' for #<Modshogun::KMeans:0x0000000188fb98> (NoMethodError)

**Comment by karlnapf** [comment]
Interesting. Can you reproduce this locally?

**Comment by OXPHOS** [comment]
I did `cmake -DRubyModular=on ..` and `make`.
And in `build/examples/meta/` I ran `make test`, I failed all ruby examples without a given reason, while all cpp examples passed.

Also I tried python modular and they failed too. Did I miss anything?

**Comment by karlnapf** [comment]
What error are they failing on?

**Comment by karlnapf** [comment]
You can also run the meta examples manually. What happens then?

**Comment by OXPHOS** [comment]
I don't think any specific info is provided. Do you mind showing me how to run meta example manually? Thanks.


Run `make test`:
```
Running tests...
Test project /Users/zora/Github/shogun/build/examples/meta
      Start  1: meta-examples-generator
 1/19 Test  #1: meta-examples-generator ...........................................   Passed    2.37 sec
      Start  2: generated_cpp-gaussian_naive_bayes
 2/19 Test  #2: generated_cpp-gaussian_naive_bayes ................................   Passed    0.09 sec
      Start  3: generated_cpp-knn
 3/19 Test  #3: generated_cpp-knn .................................................   Passed    0.05 sec
      Start  4: generated_cpp-linear_svm
 4/19 Test  #4: generated_cpp-linear_svm ..........................................   Passed    0.06 sec
      Start  5: generated_cpp-gaussian_process_regression
 5/19 Test  #5: generated_cpp-gaussian_process_regression .........................   Passed    0.21 sec
      Start  6: generated_cpp-kernel_ridge_regression
 6/19 Test  #6: generated_cpp-kernel_ridge_regression .............................   Passed    0.04 sec
      Start  7: generated_cpp-linear_ridge_regression
 7/19 Test  #7: generated_cpp-linear_ridge_regression .............................   Passed    0.04 sec
      Start  8: generated_python-classifier-gaussian_naive_bayes
 8/19 Test  #8: generated_python-classifier-gaussian_naive_bayes ..................***Failed    0.26 sec
      Start  9: generated_python-classifier-knn
 9/19 Test  #9: generated_python-classifier-knn ...................................***Failed    0.04 sec
      Start 10: generated_python-classifier-linear_svm
10/19 Test #10: generated_python-classifier-linear_svm ............................***Failed    0.03 sec
      Start 11: generated_python-gaussian_processes-gaussian_process_regression
11/19 Test #11: generated_python-gaussian_processes-gaussian_process_regression ...***Failed    0.03 sec
      Start 12: generated_python-regression-kernel_ridge_regression
12/19 Test #12: generated_python-regression-kernel_ridge_regression ...............***Failed    0.03 sec
      Start 13: generated_python-regression-linear_ridge_regression
13/19 Test #13: generated_python-regression-linear_ridge_regression ...............***Failed    0.03 sec
      Start 14: generated_ruby-classifier-gaussian_naive_bayes
14/19 Test #14: generated_ruby-classifier-gaussian_naive_bayes ....................***Failed    0.18 sec
      Start 15: generated_ruby-classifier-knn
15/19 Test #15: generated_ruby-classifier-knn .....................................***Failed    0.04 sec
      Start 16: generated_ruby-classifier-linear_svm
16/19 Test #16: generated_ruby-classifier-linear_svm ..............................***Failed    0.04 sec
      Start 17: generated_ruby-gaussian_processes-gaussian_process_regression
17/19 Test #17: generated_ruby-gaussian_processes-gaussian_process_regression .....***Failed    0.05 sec
      Start 18: generated_ruby-regression-kernel_ridge_regression
18/19 Test #18: generated_ruby-regression-kernel_ridge_regression .................***Failed    0.05 sec
      Start 19: generated_ruby-regression-linear_ridge_regression
19/19 Test #19: generated_ruby-regression-linear_ridge_regression .................***Failed    0.04 sec

37% tests passed, 12 tests failed out of 19

Total Test time (real) =   3.77 sec

The following tests FAILED:
	  8 - generated_python-classifier-gaussian_naive_bayes (Failed)
	  9 - generated_python-classifier-knn (Failed)
	 10 - generated_python-classifier-linear_svm (Failed)
	 11 - generated_python-gaussian_processes-gaussian_process_regression (Failed)
	 12 - generated_python-regression-kernel_ridge_regression (Failed)
	 13 - generated_python-regression-linear_ridge_regression (Failed)
	 14 - generated_ruby-classifier-gaussian_naive_bayes (Failed)
	 15 - generated_ruby-classifier-knn (Failed)
	 16 - generated_ruby-classifier-linear_svm (Failed)
	 17 - generated_ruby-gaussian_processes-gaussian_process_regression (Failed)
	 18 - generated_ruby-regression-kernel_ridge_regression (Failed)
	 19 - generated_ruby-regression-linear_ridge_regression (Failed)
Errors while running CTest
make: *** [test] Error 8
```

**Comment by karlnapf** [comment]
Just run the generated code listing or compiled binary from its directory

On Thursday, 19 May 2016, OXPHOS <notifications@github.com> wrote:

> I don't think any specific info is provided. Do you mind showing me how to
> run meta example manually? Thanks.
>
> Run make test:
>
> Running tests...
> Test project /Users/zora/Github/shogun/build/examples/meta
>       Start  1: meta-examples-generator
>  1/19 Test  #1: meta-examples-generator ...........................................   Passed    2.37 sec
>       Start  2: generated_cpp-gaussian_naive_bayes
>  2/19 Test  #2: generated_cpp-gaussian_naive_bayes ................................   Passed    0.09 sec
>       Start  3: generated_cpp-knn
>  3/19 Test  #3: generated_cpp-knn .................................................   Passed    0.05 sec
>       Start  4: generated_cpp-linear_svm
>  4/19 Test  #4: generated_cpp-linear_svm ..........................................   Passed    0.06 sec
>       Start  5: generated_cpp-gaussian_process_regression
>  5/19 Test  #5: generated_cpp-gaussian_process_regression .........................   Passed    0.21 sec
>       Start  6: generated_cpp-kernel_ridge_regression
>  6/19 Test  #6: generated_cpp-kernel_ridge_regression .............................   Passed    0.04 sec
>       Start  7: generated_cpp-linear_ridge_regression
>  7/19 Test  #7: generated_cpp-linear_ridge_regression .............................   Passed    0.04 sec
>       Start  8: generated_python-classifier-gaussian_naive_bayes
>  8/19 Test  #8: generated_python-classifier-gaussian_naive_bayes ..................***Failed    0.26 sec
>       Start  9: generated_python-classifier-knn
>  9/19 Test  #9: generated_python-classifier-knn ...................................***Failed    0.04 sec
>       Start 10: generated_python-classifier-linear_svm
> 10/19 Test #10: generated_python-classifier-linear_svm ............................***Failed    0.03 sec
>       Start 11: generated_python-gaussian_processes-gaussian_process_regression
> 11/19 Test #11: generated_python-gaussian_processes-gaussian_process_regression ...***Failed    0.03 sec
>       Start 12: generated_python-regression-kernel_ridge_regression
> 12/19 Test #12: generated_python-regression-kernel_ridge_regression ...............***Failed    0.03 sec
>       Start 13: generated_python-regression-linear_ridge_regression
> 13/19 Test #13: generated_python-regression-linear_ridge_regression ...............***Failed    0.03 sec
>       Start 14: generated_ruby-classifier-gaussian_naive_bayes
> 14/19 Test #14: generated_ruby-classifier-gaussian_naive_bayes ....................***Failed    0.18 sec
>       Start 15: generated_ruby-classifier-knn
> 15/19 Test #15: generated_ruby-classifier-knn .....................................***Failed    0.04 sec
>       Start 16: generated_ruby-classifier-linear_svm
> 16/19 Test #16: generated_ruby-classifier-linear_svm ..............................***Failed    0.04 sec
>       Start 17: generated_ruby-gaussian_processes-gaussian_process_regression
> 17/19 Test #17: generated_ruby-gaussian_processes-gaussian_process_regression .....***Failed    0.05 sec
>       Start 18: generated_ruby-regression-kernel_ridge_regression
> 18/19 Test #18: generated_ruby-regression-kernel_ridge_regression .................***Failed    0.05 sec
>       Start 19: generated_ruby-regression-linear_ridge_regression
> 19/19 Test #19: generated_ruby-regression-linear_ridge_regression .................***Failed    0.04 sec
>
> 37% tests passed, 12 tests failed out of 19
>
> Total Test time (real) =   3.77 sec
>
> The following tests FAILED:
>       8 - generated_python-classifier-gaussian_naive_bayes (Failed)
>       9 - generated_python-classifier-knn (Failed)
>      10 - generated_python-classifier-linear_svm (Failed)
>      11 - generated_python-gaussian_processes-gaussian_process_regression (Failed)
>      12 - generated_python-regression-kernel_ridge_regression (Failed)
>      13 - generated_python-regression-linear_ridge_regression (Failed)
>      14 - generated_ruby-classifier-gaussian_naive_bayes (Failed)
>      15 - generated_ruby-classifier-knn (Failed)
>      16 - generated_ruby-classifier-linear_svm (Failed)
>      17 - generated_ruby-gaussian_processes-gaussian_process_regression (Failed)
>      18 - generated_ruby-regression-kernel_ridge_regression (Failed)
>      19 - generated_ruby-regression-linear_ridge_regression (Failed)
> Errors while running CTest
> make: *** [test] Error 8
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/3183#issuecomment-220364653>
>


-- 
Sent from my phone

**Comment by OXPHOS** [comment]
Aha @karlnapf If this is what you refer to - `$ruby build/examples/meta/python/classifier/knn.rb `
I have the same `modshogun` issue
```
l_require.rb:55:in `require': cannot load such file -- modshogun (LoadError)
	from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from build/examples/meta/ruby/classifier/knn.rb:1:in `<main>'
```

**Comment by karlnapf** [comment]
That is weird. Is it the same error as you get on travis?
Also, your Python examples also seem to fail, why is that? Are you sure your build is working fine?

**Comment by OXPHOS** [comment]
@karlnapf So I tried both make install and add path.
Python will crash every time I try to run any example (real crash kind of crash)
Ruby returns the same error as on travis for `kmeans.rb`

```
Mac148934:classifier zora$ ruby kmeans.rb 
/Library/Ruby/Site/2.0.0/universal-darwin14/modshogun.bundle: warning: rb_define_const: invalid name `__MAC_10_8' for constant
kmeans.rb:28:in `<main>': undefined method `set_mbKMeans_params' for #<Modshogun::KMeans:0x007fbc051695e0> (NoMethodError)
```

**Comment by karlnapf** [comment]
Can we discuss this in another thread?

**Comment by OXPHOS** [comment]
@karlnapf This should be moved to clustering as well. What do you think of the ruby issue? You want to open a new thread for it?

**Comment by karlnapf** [comment]
Yes move it!
Is the current verison in the PR producing the problems you mentioend? Then I will close locally and check

**Comment by karlnapf** [comment]
Ill just use this version to check now

**Comment by OXPHOS** [comment]
@karlnapf  yes this is the version gives problems. thx.

**Comment by OXPHOS** [comment]
I tried rebasing but it claimed there's conflict I cannot solve..So this way won't work. I'll change it back.

However at local there's a new issue **similar** to the qda one for ruby, with `kmeans.sg` line26: `kmeans_mb.set_mini_batch_size(3)`.


Whenever I tried to set or get `batch_size` or `minib_iter`, and `ruby kmeans.rb`, I got:
```
/Library/Ruby/Site/2.0.0/universal-darwin14/modshogun.bundle: warning: rb_define_const: invalid name `__MAC_10_8' for constant
[ERROR] In file /Users/zora/Github/shogun/src/shogun/io/File.cpp line 71: Error opening file '../../data/classifier_binary_2d_linear_features_train.dat'
ruby/clustering/kmeans.rb:3:in `initialize': [ERROR] In file /Users/zora/Github/shogun/src/shogun/io/File.cpp line 71: Error opening file '../../data/classifier_binary_2d_linear_features_train.dat' (fatal)
	from ruby/clustering/kmeans.rb:3:in `new'
	from ruby/clustering/kmeans.rb:3:in `<main>'
```

I can do `set_k()`, but cannot set or get anything related to mini_batch (i.e. `batch_size` and `minib_iter`). The two variables are declared and defined at instantiation.

File: src/shogun/clustering/KMeans.cpp [hunk 7]
```diff
@@ -75,9 +75,9 @@ void CKMeans::set_initial_centers(SGMatrix<float64_t> centers)
  	CDenseFeatures<float64_t>* lhs=((CDenseFeatures<float64_t>*) distance->get_lhs());
 	dimensions=lhs->get_num_features();
 	REQUIRE(centers.num_cols == k,
-			"Expected %d initial cluster centers, got %d", k, centers.num_cols);
+			"Expected %d initial cluster centers, got %d.\n", k, centers.num_cols);
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: you did not rebase against develop -- this is my patch

**Comment by karlnapf** [comment]
All these things cannot really happen ;)
I suggest you clean up your branch and then we see how it goes ... I manually added your kmeans.sg locally and it works

**Comment by karlnapf** [comment]
Travis also seems very content

**Comment by karlnapf** [comment]
And it is indeed good that you removed the kmeans examples from the other example dir

**Comment by OXPHOS** [comment]
...This is weird..I'll try it again at local. And rebase. Thx!

**Comment by karlnapf** [comment]
This is great and ready to go. We just need some integration testing data now.

**Comment by OXPHOS** [comment]
@karlnapf  So here's the problem - Kmeans currently sets random centers, so the order of the centers might be different. If the data is split into two clusters, only 50% chance it will match the dataset. I tried `set_fixed_center()` but it doesn't seem to solve the problem completely (10%+ I got different results, and 50% the 15th decimal digits are different).

To set the centers explicitly, SGMatrix is required. I am not sure whether this can be done.
Is providing multiple data file an option?
Or can we add an argument in kmeans to set initial centers to (0, 0)?

**Comment by karlnapf** [comment]
One way to fix this is to fix seeds, we have to decide whether we want to do this or not.
Just talked to @lisitsyn and he is up for doing it.

So let us experiment here. Can you initialise the seed, update the data. Then it should be fixed

@sanuj @Saurabh7 @arianepaola this also goes for your cookbook examples (GMM can have some data now as well)

**Comment by karlnapf** [comment]
http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CMath.html#a4a93b2cd4ca6d7773e15f965501a856e

**Comment by OXPHOS** [comment]
@karlnapf 
`Math.init_random(1)` won't work, because it is converted to Math->init_random(1) in cpp.
`Math::init_random(1)` won't work, because `::` is unknown.

Also, I have `0` in `t_NUMERAL` in `parse.py` but I still cannot use 0.
And last,  Math.init_random(0) seems to generate random seed?
```
if (initseed==0)
	seed = CRandom::generate_seed();
else
	seed=initseed;
```
So still not solved..

**Comment by vigsterkr** [comment]
@OXPHOS `sg_rand->set_seed(1)`, but what is actually the problem with `CMath::init_random(1)`? it's a static function you should be able to call it like that

**Comment by OXPHOS** [comment]
@vigsterkr Thanks. Meta example language doesn't allow `:` literal

**Comment by OXPHOS** [comment]
`->` won't work either

**Comment by karlnapf** [comment]
Single colon `Math:init_random(0)`

Let me know if that still fails.

**Comment by OXPHOS** [comment]
@karlnapf  It doesn't work for me at local..

**Comment by karlnapf** [comment]
Can it be that your branch is not up to date? For me
`make meta examples`
`make build_cpp_meta_examples`
and executing `build/examples/meta/cpp/classifier/kmeans` (from its directory)
all work with the latest `kmeans.sg` from this PR (even with using 0 as seed)
`make test` also works for the Python example

Travis also says they work (cpp). It is just the integration test doesnt work, which probably is since the data file is wrong. 

Python doesnt work due to a bug (see other email, we are fixing it)

**Comment by OXPHOS** [comment]
My branch is up-to-date. And the `t_NUMERAL` is the `t_NUMERAL = "[0-9]+(\.[0-9]+)?"` as you have updated in `parse.py`.
When I try to `make build_cpp_meta_examples` it raises error for illegal literal..
Is there anything I should completely delete/reset?

I did `rm -rf build` and `make clean`, not helping..

**Comment by OXPHOS** [comment]
I did a little trick:
```
Math math()
math.init_random(1)
```
Now I can generate consistent data file. If you don't mind setting the seed = 1, I can just send in the integration test dataset generated with `Math:init_random(1)` for now.

But I do have problem working with `:` and `0`..

**Comment by sorig** [comment]
I'm experiencing the same parsing issue with static calls. Looking into it now. Will keep you updated.

**Comment by karlnapf** [comment]
Just merged #3241 so the static calls should work now. Cleaner.

Make sure to remove
```
examples/meta/parsetab.py
examples/meta/lextab.py```
otherwise there might be weird errors

Can you rebase, make sure the integration test works, and then ping me?

**Comment by OXPHOS** [comment]
@karlnapf it is failing on current kmeans integration test I guess..could you merge the updated dataset? thanks!

**Comment by karlnapf** [comment]
Let me know

**Comment by OXPHOS** [comment]
@karlnapf @sanuj 
make error for java:
```
/Users/zora/Github/shogun/build/examples/meta/java/clustering/kmeans.java:16: error: reference to Math is ambiguous
Math.init_random(0);
^
  both class org.shogun.Math in org.shogun and class java.lang.Math in java.lang match
1 error
make[2]: *** [examples/meta/java/compiled_java_clustering-kmeans] Error 1
make[1]: *** [examples/meta/java/CMakeFiles/build_java_meta_examples.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by OXPHOS** [comment]
For R:
```
Welcome! This is SHOGUN version 4.2.0
Error: unexpected '$' in "Math$$"
Execution halted
```

**Comment by OXPHOS** [comment]
This one seems to be `init_random()` independent:

Ruby:
```
30/40 Test #30: generated_ruby-clustering-kmeans ................................***Failed    0.09 sec
/opt/shogun/build/examples/meta/ruby/clustering/kmeans.rb:4:in `<main>': undefined method `type' for Modshogun:Module (NoMethodError)
```

**Comment by karlnapf** [comment]
ok so the java error should be resolved in #3251

**Comment by karlnapf** [comment]
The ruby one is fixed in ae0e85aef2c97c88c77453e80cfb0b3379c140e4 (sorry I made a mistake earlier)

**Comment by karlnapf** [comment]
And the one in R is another copy paste error or mine.
Fixed and pushed in the second commit of #3251 
Need to wait for travis, then can merge that. Ping me if you see it is fine

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 8]
```diff
@@ -0,0 +1,51 @@
+==================
+K-means clustering
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: remove "clustering"

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+:sgclass:`CKMeans` also supports mini batch :math:`k`-means clustering.
+We can create an instance of :sgclass:`CKMeans` classifier with mini batch :math:`k`-means method by providing the batch size and iteration number.
+
+.. sgexample:: kmeans.sg:create_instance_mb
+
+Then train the dataset and extract the centers and radius information as mentioned above.
+
+----------
+References
+----------
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: can you add david barbers book as reference as well?

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 10]
```diff
@@ -0,0 +1,51 @@
+==================
+K-means clustering
+==================
+:math:`K`-means clustering aims to partition :math:`n` observations into :math:`k\leq n` clusters (sets :math:`\mathbf{S}`),
+in which each observation belongs to the cluster with the nearest mean, serving as a prototype of the cluster.
+
+In other words, its objective is to find:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: is to minimize:

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+.. math::
+   k = \argmin_\mathbf{S} \sum_{i=1}^{k}\sum_{\mathbf{x}\in S_k}\left \|\boldsymbol{x} - \boldsymbol{\mu}_i  \right \|^{2}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: remove k= at the beginning

**Comment by karlnapf** [comment]
Ok travis failure is from gmm. As soon as the minor adjustments in the rst are made, we can (finally!) merge :)

**Comment by OXPHOS** [comment]
@karlnapf can check again :)

**Comment by OXPHOS** [comment]
The R is also failing..

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 12]
```diff
@@ -0,0 +1,56 @@
+=======
+K-means
+=======
+:math:`K`-means clustering aims to partition :math:`n` observations into :math:`k\leq n` clusters (sets :math:`\mathbf{S}`),
+in which each observation belongs to the cluster with the nearest mean, serving as a prototype of the cluster.
+
+In other words, its objective is to minimize:
+
+.. math::
+   x\argmin_\mathbf{S} \sum_{i=1}^{k}\sum_{\mathbf{x}\in S_k}\left \|\boldsymbol{x} - \boldsymbol{\mu}_i  \right \|^{2}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: this x in the beginning should not be there :)

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+.. sgexample:: kmeans.sg:create_features
+
+In order to run :sgclass:`CKMeans`, we need to choose a distance, for example :sgclass:`CEuclideanDistance`, or other sub-classes of :sgclass:`CDistance`. The distance is initialized with the data we want to classify.
+
+.. sgexample:: kmeans.sg:choose_distance
+
+Once we have chosen a distance, we create an instance of the :sgclass:`CKMeans` classifier.
+We explicitly set :math:`k`, the number of clusters we are expecting to have as 3 and pass it to :sgclass:`CKMeans`. In this example, we apply Lloyd's method for `k`-means clustering.
+
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: maybe mention (or remove in code) this "fixed_centers"?

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+.. sgexample:: kmeans.sg:create_instance_lloyd
+
+Then we train the dataset:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: the model, not the dataset

File: doc/cookbook/source/examples/clustering/kmeans.rst [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+.. sgexample:: kmeans.sg:train_dataset
+
+And we can extract centers and radius of each cluster:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/kmeans.rst]: Dont start sentences with "And"

**Comment by karlnapf** [comment]
Some minor adjustments are needed. A rebase against develop should make travis happy (I disabled R)

**Comment by OXPHOS** [comment]
@karlnapf I tried ~30 times on local and the test can be passed, but it keeps failing on travis..So I'll put `fixed_centers()` back for now but I don't think we should show that in the cookbook.

**Comment by karlnapf** [comment]
fixed centers obviously changes the algorithm (and the integration test data)
The point I was making: do we need that? If so, you should mention what it does.
Though I preferred if you changed the integration test data and removed it

**Comment by OXPHOS** [comment]
@karlnapf I updated the integration test https://github.com/shogun-toolbox/shogun-data/pull/101
I'll update the meta example once the data is merged. thx!

**Comment by karlnapf** [comment]
Merged :)

**Comment by karlnapf** [comment]
Next time you squash: please put a meaningful commit message.
"Add kmeans cookbook and meta example"

**Comment by karlnapf** [comment]
Waiting for travis and then merging

**Comment by OXPHOS** [comment]
```
[ 98%] Building CXX object examples/meta/cpp/CMakeFiles/kmeans.dir/clustering/kmeans.cpp.o
/opt/shogun/build/examples/meta/cpp/clustering/kmeans.cpp:31:42: error: use of undeclared identifier 'KMM_LLOYD'
auto kmeans = some<CKMeans>(2, distance, KMM_LLOYD);
                                         ^
/opt/shogun/build/examples/meta/cpp/clustering/kmeans.cpp:44:45: error: use of undeclared identifier 'KMM_MINI_BATCH'
auto kmeans_mb = some<CKMeans>(2, distance, KMM_MINI_BATCH);
                                            ^
2 errors generated.
make[2]: *** [examples/meta/cpp/CMakeFiles/kmeans.dir/clustering/kmeans.cpp.o] Error 1
make[1]: *** [examples/meta/cpp/CMakeFiles/kmeans.dir/all] Error 2
make: *** [all] Error 2
```
?????
It works for me at local..

**Comment by karlnapf** [comment]
Aaaand merging

**Pull request merged**